h/traversal/group_membership.py

```diff
@@ -13,6 +13,13 @@ class GroupMembershipContext:
     membership: GroupMembership | None
 
 
+@dataclass
+class AddGroupMembershipContext:
+    group: Group
+    user: User
+    roles: list[GroupMembershipRoles]
```

Comment:
Pyramid has the very handy concept of a "context object" for a request that gets passed into various different parts of the request handling code. For example the permissions predicate functions in [`predicates.py`](https://github.com/hypothesis/h/blob/b64c396143c142adb62c08a4df4bf3a01ea30746/h/security/predicates.py) receive `context` as an argument. And so do the view functions in [`views/api/group_members.py`](https://github.com/hypothesis/h/blob/b64c396143c142adb62c08a4df4bf3a01ea30746/h/views/api/group_members.py).

The `context` object is a holder for things that're useful for different layers of code handling the request. In this case:

* `AddGroupMembershipContext.group` is the group that the request is trying to add a membership to
* `AddGroupMembershipContext.user` is the user that the request is trying to add to the group
* `AddGroupMembershipContext.roles` is the roles that the request is trying to add the user with

The way this works is:

1. The [route configuration for the `"api.group_member"` route](https://github.com/hypothesis/h/blob/b64c396143c142adb62c08a4df4bf3a01ea30746/h/routes.py#L159-L163) declares `group_membership_api_factory()` as the "factory" 
2.